### ISSUE TITLE
False hits in test coverage from gcov

### ISSUE THREAD

**Issue by gilles-peskine-arm** [Mbed-TLS/mbedtls#6958]
It appears that our test coverage reports are overly optimistic: they report certain lines as hit, but the lines are actually not hit. We initially thought this was an [issue in lcov](https://github.com/linux-test-project/lcov/issues/185) (the coverage reporting tool), but it's actually due to `gcov` (the coverage measurement tool), which is part of the GCC toolchain.

It appears that clang/LLVM works better. @bensze01 writes:

> First, apply this patch:
> ```
> diff --git a/scripts/gcov_tool.sh b/scripts/gcov_tool.sh
> new file mode 100755
> index 000000000..7c663d8da
> --- /dev/null
> +++ b/scripts/gcov_tool.sh
> @@ -0,0 +1,2 @@
> +#!/bin/sh
> +exec ${GCOV_TOOL:-gcov} "$@"
> diff --git a/scripts/lcov.sh b/scripts/lcov.sh
> index 8d141eedf..02e9c25ec 100755
> --- a/scripts/lcov.sh
> +++ b/scripts/lcov.sh
> @@ -46,10 +46,10 @@ set -eu
>  lcov_library_report () {
>      rm -rf Coverage
>      mkdir Coverage Coverage/tmp
> -    lcov --capture --initial --directory library -o Coverage/tmp/files.info
> -    lcov --rc lcov_branch_coverage=1 --capture --directory library -o Coverage/tmp/tests.info
> -    lcov --rc lcov_branch_coverage=1 --add-tracefile Coverage/tmp/files.info --add-tracefile Coverage/tmp/tests.info -o Coverage/tmp/all.info
> -    lcov --rc lcov_branch_coverage=1 --remove Coverage/tmp/all.info -o Coverage/tmp/final.info '*.h'
> +    lcov --gcov-tool "$PWD/scripts/gcov_tool.sh" --capture --initial --directory library -o Coverage/tmp/files.info
> +    lcov --gcov-tool "$PWD/scripts/gcov_tool.sh" --rc lcov_branch_coverage=1 --capture --directory library -o Coverage/tmp/tests.info
> +    lcov --gcov-tool "$PWD/scripts/gcov_tool.sh" --rc lcov_branch_coverage=1 --add-tracefile Coverage/tmp/files.info --add-tracefile Coverage/tmp/tests.info -o Coverage/tmp/all.info
> +    lcov --gcov-tool "$PWD/scripts/gcov_tool.sh" --rc lcov_branch_coverage=1 --remove Coverage/tmp/all.info -o Coverage/tmp/final.info '*.h'
>      gendesc tests/Descriptions.txt -o Coverage/tmp/descriptions
>      genhtml --title "mbed TLS" --description-file Coverage/tmp/descriptions --keep-descriptions --legend --branch-coverage -o Coverage Coverage/tmp/final.info
>      rm -f Coverage/tmp/*.info Coverage/tmp/descriptions
> ```
> 
> Then build the coverage data like this:
> ```
> make CC=clang-10 CFLAGS='--coverage -g3 -O0' LDFLAGS='--coverage' -C tests test_suite_x509parse
> (cd tests && ./test_suite_x509parse)
> make lcov GCOV_TOOL="llvm-cov-10 gcov"
> (replace clang-10 and llvm-cov-10 with the version you have installed)
> ```
> 

The goal of this task is to investigate the coverage measurement differences between different versions of GCC and LLVM. In particular:

* Is this a bug that is fixed in the latest GCC?
* Is this a known bug in GCC?
* Where there are differences, is LLVM always correct?

If LLVM is more reliable than GCC, we'll switch our scripts to use it.

**Comment by AndrzejKurek** [comment]
I don't think this will be an s-sized task. Basic comparison for just one test case from:
- [here for gcov](https://github.com/AndrzejKurek/mbedtls/commit/4fc88367f0bfb67753d51781c03330788bdedc28) (updated version mentioned in https://github.com/linux-test-project/lcov/issues/185), vs
- [here for llvm-cov](https://github.com/AndrzejKurek/mbedtls/commit/9fd6cea4cdd758cc476ca26302eb49ed1fe51d86)

Shows that llvm-cov also shows some faulty behaviour:
- it shows coverage for line 789 of x509.c, while the line isn't executed. gcov coverage works fine in this particular case:
![gcov](https://user-images.githubusercontent.com/29195988/214825705-edc4228f-70d7-4169-9fbf-5a068e926087.png)
![llvm-cov](https://user-images.githubusercontent.com/29195988/214825717-68703594-784a-47da-b4ed-36bfb0e68e06.png)

Lcov using gcov on the other hand has some workarounds that seem to work fine (in the limited scope I investigated), but I haven't studied the impact on a bigger scope.

Run commands:
gcov version:
```
make CFLAGS='--coverage -g3 -O0' LDFLAGS=--coverage -C tests test_suite_x509parse 
(cd tests && ./test_suite_x509parse)
make lcov
firefox Coverage/index.html
```
llvm-cov version:
```
make CC=clang-14 CFLAGS='--coverage -g3 -O0' LDFLAGS='--coverage' -C tests test_suite_x509parse
(cd tests && ./test_suite_x509parse)
make lcov GCOV_TOOL="llvm-cov-14 gcov"
```
Results of coverage runs (and two diffs, llvm-cov vs updated gcov is what interests us):
[coverage_tests.tar.gz](https://github.com/Mbed-TLS/mbedtls/files/10508769/coverage_tests.tar.gz)

Please note that the diff's LBC might also be inconsistent, as shown [here](https://github.com/linux-test-project/lcov/issues/185#issuecomment-1404872972).

It seems that for some reason it's hard to get flawless results with our code. Bugs in tools? Any options that we don't use, and we should? I don't have the answers for now.
